### ISSUE TITLE
SC2154: Doesn't follow simple indirection on things like `read`

### ISSUE THREAD

**Issue by LukeShu** [koalaman/shellcheck#1017]
#### For bugs
- Rule Id: SC2154
- My shellcheck version: 0.4.6
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

# It knows that `read` sets a variable
read -r -a var1 <<<'foo bar'
printf ' -> %s\n' "${var1[@]}"

# But not if we add a layer of indirection
set_array() {
	read -r -a "$1" <<<$"2"
}
set_array var2 'foo bar'
printf ' -> %s\n' "${var2[@]}"
```

#### Here's what shellcheck currently says:

```
 
Line 12:
printf ' -> %s\n' "${var2[@]}"
                   ^-- SC2154: var2 is referenced but not assigned (did you mean 'var1'?).

```

#### Here's what I wanted or expected to see:

Nothing, it should come back clean.

**Comment by koalaman** [comment]
You're right of course. It's a ShellCheck's design decision not to try to figure out whether variables are used through indirection. Doing that is an infinite rabbit hole that in the general case is equivalent to the halting problem. 

While trying to add this simple case might work and provide some value, you'd quickly find trivial additions that break it again (like `read -r -a "${1%?}"`). Instead, ShellCheck just categorically stops at the indirection boundary.

Sorry about that. Consider explicitly assigning the variables you know will be used or disabling this warning script-wide instead.

**Comment by LukeShu** [comment]
The simple case would have made me happy, but I understand categorically avoiding indirection.  In that case, would you be willing to add a directive a directive that indicates that a function sets a variable whose name is the n-th argument, or similar?  I might be getting in to an XY problem...
